HCR: Can not link exported const, in external library

I'm trying to use this library: https://github.com/Vindaar/shell
  with Nim HCR, and LD linker can't link exported const symbol, for some reason:


  This is my fork of this library, with added export symbol (original didn't have):
    [https://github.com/xtazy-drop/test-nim-hcr-shell](https://github.com/xtazy-drop/test-nim-hcr-shell)

```
    ...
    
    const defaultDebugConfig*: set[DebugOutputKind] =
      block:
        var config: set[DebugOutputKind] = {
          dokOutput, dokError, dokCommand, dokRuntime
        }
        
    ...
```     
        
### Compilation output:
  
```
$ nim c --outdir:f --hotcodereloading:on  --hints:off test_nim_hcr_shell.nim

/usr/bin/ld: /home/ubuntu-mate/.cache/nim/test_nim_hcr_shell_d/@mtest_nim_hcr_shell.nim.c.o: in function `NimMainModule':
@mtest_nim_hcr_shell.nim.c:(.text+0xec8): undefined reference to `defaultDebugConfig__LgOdL9aGyblIQIS067yE8vw'
collect2: error: ld returned 1 exit status
```

  When I link manually, against .so file of this library, then it links fine (nim v1.0.6):
    ```
    $ gcc   -o f/test_nim_hcr_shell $cache_dir/@mtest_nim_hcr_shell.nim.c.o  $cache_dir/lib@mf@sshell@sshell.nim.c.so   -lm -lrt   -ldl
    ```
      - works for nim v1.0.6
      - doesn't work for nim v1.2.0


### Test script:
  
  ```
  $ git clone https://github.com/xtazy-drop/test-nim-hcr-shell
  $ cd test-nim-hcr-shell/
  $ ./get-shell.sh
  $ ./compile.sh
  ```
  

### defaultDebugConfig_* symbol generations: 

  Nim v1.0.6:
    Generated file ```@mf@sshell@sshell.nim.c```, line 230:
```
NIM_CONST tySet_tyEnum_DebugOutputKind__RLJKb7vjqsoNJNVirVUgNg defaultDebugConfig__LgOdL9aGyblIQIS067yE8vw = 15;
```

  Symbol record in .so file:
```
   $ objdump -t /home/ubuntu-mate/.cache/nim/test_nim_hcr_shell_d/lib@mf@sshell@sshell.nim.c.so | grep defaultDebugConfig
      00000000000031e0 g     O .rodata	0000000000000001              defaultDebugConfig__LgOdL9aGyblIQIS067yE8vw
```
      
  Nim v1.2.0:
    Generated file @mf@sshell@sshell.nim.c, line 194:
```
N_LIB_PRIVATE NIM_CONST tySet_tyEnum_DebugOutputKind__RLJKb7vjqsoNJNVirVUgNg defaultDebugConfig__LgOdL9aGyblIQIS067yE8vw = 15;
```

  Symbol record in .so file::
```
  $ objdump -t ~/.cache/nim/test_nim_hcr_shell_d/lib@mf@sshell@sshell.nim.c.so | grep defaultDebugConfig
      00000000000031e0 l     O .rodata	0000000000000001              defaultDebugConfig__LgOdL9aGyblIQIS067yE8vw
```

--- a/compiler/cgen.nim
+++ b/compiler/cgen.nim
@@ -1166,21 +1166,43 @@ proc requestConstImpl(p: BProc, sym: PSym) =
   useHeader(m, sym)
   if sym.loc.k == locNone:
     fillLoc(sym.loc, locData, sym.ast, mangleName(p.module, sym), OnStatic)
+  if m.hcrOn: incl(sym.loc.flags, lfIndirect)
+
   if lfNoDecl in sym.loc.flags: return
   # declare implementation:
   var q = findPendingModule(m, sym)
   if q != nil and not containsOrIncl(q.declaredThings, sym.id):
     assert q.initProc.module == q
+    # add a suffix for hcr - will later init the global pointer with this data
+    let actualConstName = if m.hcrOn: sym.loc.r & "_const" else: sym.loc.r
     q.s[cfsData].addf("N_LIB_PRIVATE NIM_CONST $1 $2 = $3;$n",
-        [getTypeDesc(q, sym.typ), sym.loc.r, genBracedInit(q.initProc, sym.ast, isConst = true)])
+        [getTypeDesc(q, sym.typ), actualConstName, genBracedInit(q.initProc, sym.ast, isConst = true)])
+    if m.hcrOn:
+      # generate the global pointer with the real name
+      q.s[cfsVars].addf("static $1* $2;$n", [getTypeDesc(m, sym.loc.t), sym.loc.r])
+      # register it (but ignore the boolean result of hcrRegisterGlobal)
+      q.initProc.procSec(cpsLocals).addf(
+        "\thcrRegisterGlobal($1, \"$2\", sizeof($3), NULL, (void**)&$2);$n",
+        [getModuleDllPath(q, sym), sym.loc.r, rdLoc(sym.loc)])
+      # always copy over the contents of the actual constant with the _const
+      # suffix ==> this means that the constant is reloadable & updatable!
+      q.initProc.procSec(cpsLocals).add(ropecg(q,
+        "\t#nimCopyMem((void*)$1, (NIM_CONST void*)&$2, sizeof($3));$n",
+        [sym.loc.r, actualConstName, rdLoc(sym.loc)]))
   # declare header:
   if q != m and not containsOrIncl(m.declaredThings, sym.id):
     assert(sym.loc.r != nil)
-    let headerDecl = "extern NIM_CONST $1 $2;$n" %
-        [getTypeDesc(m, sym.loc.t), sym.loc.r]
-    m.s[cfsData].add(headerDecl)
-    if sfExportc in sym.flags and p.module.g.generatedHeader != nil:
-      p.module.g.generatedHeader.s[cfsData].add(headerDecl)
+    if m.hcrOn:
+      m.s[cfsVars].addf("static $1* $2;$n", [getTypeDesc(m, sym.loc.t), sym.loc.r]);
+      m.initProc.procSec(cpsLocals).addf(
+        "\t$1 = ($2*)hcrGetGlobal($3, \"$1\");$n", [sym.loc.r,
+        getTypeDesc(m, sym.loc.t), getModuleDllPath(q, sym)])
+    else:
+      let headerDecl = "extern NIM_CONST $1 $2;$n" %
+          [getTypeDesc(m, sym.loc.t), sym.loc.r]
+      m.s[cfsData].add(headerDecl)
+      if sfExportc in sym.flags and p.module.g.generatedHeader != nil:
+        p.module.g.generatedHeader.s[cfsData].add(headerDecl)
 
 proc isActivated(prc: PSym): bool = prc.typ != nil
 

--- a/tests/dll/nimhcr_0_3.nim
+++ b/tests/dll/nimhcr_0_3.nim
@@ -14,5 +14,6 @@ let c = makeCounter()
 afterCodeReload:
   echo "   0: after - closure iterator: ", c()
   echo "   0: after - closure iterator: ", c()
+  echo "   0: after - c_2 = ", c_2
 
 proc getInt*(): int = return g_1 + g_2.len

--- a/tests/dll/nimhcr_2_1.nim
+++ b/tests/dll/nimhcr_2_1.nim
@@ -7,6 +7,8 @@ type
 
 let g_2* = @[Type2(data: 2), Type2(data: 3)][1..^1] # should have a length of 1
 
+const c_2* = [1, 2, 3] # testing that a complext const object is properly exported
+
 var a: tuple[str: string, i: int]
 a.str = "   2: random string"
 echo a.str

--- a/tests/dll/nimhcr_integration.nim
+++ b/tests/dll/nimhcr_integration.nim
@@ -36,6 +36,7 @@ max mutual recursion reached!
 bar
    0: after - closure iterator: 0
    0: after - closure iterator: 1
+   0: after - c_2 = [1, 2, 3]
 main: after
               The answer is: 9
 main: hasAnyModuleChanged? true

